compilers/concrete-compiler/compiler/lib/Runtime/GPUDFG.cpp

```diff
@@ -1064,15 +1074,24 @@ void memref_keyswitch_lwe_u64_process(Process *p, int32_t loc, int32_t chunk_id,
       return dep;
     } else {
       // Schedule the keyswitch kernel on the GPU
-      cudaStream_t *s = (cudaStream_t *)p->dfg->get_gpu_stream(loc);
+      cudaStream_t s = (cudaStream_t)p->dfg->get_gpu_stream(loc);
       void *ct0_gpu = d->device_data;
       void *out_gpu = cuda_malloc_async(data_size, s, loc);
       void *ksk_gpu = p->ctx.val->get_ksk_gpu(
           p->level.val, p->input_lwe_dim.val, p->output_lwe_dim.val, loc, s,
           p->sk_index.val);
+        // Initialize indexes
+      uint64_t *indexes = (uint64_t*) malloc(num_samples * sizeof(uint64_t));
+      for (uint32_t i = 0; i < num_samples; i++) {
+	indexes[i] = i;
+      }
+      void *indexes_gpu = alloc_and_memcpy_async_to_gpu(indexes, 0, num_samples, loc, s);
+
       cuda_keyswitch_lwe_ciphertext_vector_64(
-          s, loc, out_gpu, ct0_gpu, ksk_gpu, p->input_lwe_dim.val,
+	  s, loc, out_gpu, indexes_gpu, ct0_gpu, indexes_gpu, ksk_gpu, p->input_lwe_dim.val,
           p->output_lwe_dim.val, p->base_log.val, p->level.val, num_samples);
+      cuda_drop_async(indexes_gpu, s, loc);
+      free(indexes);
```

Comment:
If this is freed here, the CUDA runtime will likely try reading this array much later, once it can actually do the memory transfer. This needs to be deallocated after a synchronization point - see below.